text_generation/causal_lm/cpp/prompt_lookup_decoding_lm.cpp

```diff
@@ -0,0 +1,299 @@
+    model.get_tensor("beam_idx").set_shape({BATCH_SIZE});
+    model.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+
+    // To coollect kv-cache for the <PROMPT> and to get the next token run the very first infer request
+    model.infer();
+
+    // logits shape is [BATCH_SIZE, seq_len, vocab_size]
+    auto logits = model.get_tensor("logits");
+    size_t vocab_size = logits.get_shape().back();
+    auto data_logits = logits.data<float>() + (seq_len - 1) * vocab_size;
+    int64_t out_token = std::max_element(data_logits, data_logits + vocab_size) - data_logits;
+
+    full_input_ids.push_back(out_token);
+
+    auto first_token = out_token;
+    text_streamer.put(out_token);
+
+    const int64_t EOS_TOKEN = get_eos_token(tokenizer_model);
+
+    // Prompt lookup decoding is a speculative decoding technic where the draft model replaced
+    // with string matching in the prompt to generate candidate token sequences.
+    int max_sequence_length = 100;
+    PromptLookupCandidateGenerator candidateGenerator{3, 5};
+
+    while (out_token != EOS_TOKEN && seq_len < max_sequence_length) {
+        auto candidates = candidateGenerator.generate_candidates(full_input_ids);
+        size_t candidates_size = candidates.size();
+
+        // candidates_size + 1 tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, candidates_size + 1});
+        input_ids.data<int64_t>()[0] = first_token;
+        for (int i = 0; i < candidates_size; i++) {
+            input_ids.data<int64_t>()[i + 1] = candidates[i];
+        }
+
+        attention_mask.set_shape({BATCH_SIZE, seq_len + candidates_size + 1});
+        std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
+
+        position_ids.set_shape({BATCH_SIZE, candidates_size + 1});
+        std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), seq_len);
+
+        model.infer();
+
+        data_logits = logits.data<float>();  // [BATCH_SIZE, 1 + candidates_size, vocab_size]
+
+        // Find candidates matches.
+        // Accept first token (if not EOS) as it generated from the valid input_ids sequence.
+        // Iterate only if canidate token matches model predicted token.
+        // Accept last token as it generated from the valid predicted tokens sequence.
+        size_t accepted_tokens_number = 0;
+        for (size_t i = 0; i < candidates_size + 1; i++) {
+            auto start = data_logits + vocab_size * i;
+            auto stop = data_logits + vocab_size * (i + 1);
+            auto candidate_out_token = std::max_element(start, stop) - start;
+            out_token = candidate_out_token;
+
+            if (candidate_out_token == EOS_TOKEN) {
+                break;
+            }
+
+            text_streamer.put(candidate_out_token);
+            full_input_ids.push_back(candidate_out_token);
+            accepted_tokens_number++;
```

Comment:
should these 3 lines be after condition below?

Reply:
It seems correct to as it is right now. I'll answer in more detail in next comment